editor/src/components/canvas/controls/outline-control.tsx

```diff
@@ -58,9 +56,9 @@ function isDraggingToMove(
   return dragState != null && dragState?.type === 'MOVE_DRAG_STATE' && targetIsDragged
 }
 
-export class OutlineControls extends React.Component<OutlineControlsProps> {
-  getDragStateFrame = (target: TemplatePath): CanvasRectangle | null => {
-    const { dragState } = this.props
+export const OutlineControls = (props: OutlineControlsProps) => {
+  const getDragStateFrame = (target: TemplatePath): CanvasRectangle | null => {
```

Comment:
this is a good idea, converting an old class component to a function component! :) 

these functions that used to live on the Class are now turned into inline functions. there's some important differences between how the two kinds work – in the previous version, the `getDragStateFrame` function would be instantiated once when the class lifecycle begins, and we would use the same reference for every new call of the render() function. In this new function component, for every single render this line of code will re-run, creating a new instance of the function `getDragStateFrame`. The most important thing to know is that this means `getDragStateFrame` will have a new _reference_ for every render. Because a lot of React depends on the reference being stable (being the same from render to render), this can very easily lead to performance problems.

The easy solution is to just remember, as a rule, **when you are doing a class conversion, you should always wrap the functions to a `React.useCallback`**, like so:

```javascript
const getDragStateFrame = React.useCallback((target: TemplatePath): CanvasRectangle | null => {
   ....
  }, 
[   /* list the dependencies here */ ])
```

This will bring the behavior of the functions a little bit closer to how it worked in the class!

Reply:
Okay, I think this needs futher elaboration. 

There are many exceptions to this rule, a function actually only needs React.useCallback() wrapping if it is defined inside the Function component and
1. used as a prop in the JSX 
2. listed as a dependency of a hook (like, for example, another useCallback is listing this function as a dependency)

But because it is hard to tell when this rule should be appplied (our `eslint` setup struggles with detecting some mistakes too) I would advise that it is best to just stick to it and wrap everything in a React.useCallback when you are converting a class to a function component 🙂

BTW, when you are creating a new component, it is best if you don't make functions inside other functions unless you actually need to use them as callbacks. But because this is code conversion, it's perfectly acceptable that you did not refactor this function to be in the root level of the file (which would be the nicest, code-quality-wise)

Reply:
Commit 3420f3d addresses this, thanks for the feedback!